obs-studio-server/source/memory-manager.cpp

```diff
@@ -148,218 +155,171 @@ void MemoryManager::addCachedMemory(source_info *si)
-	current_cached_size += si->size;
-	si->cached = true;
+	blog(LOG_INFO, "adding %dMB, source: %s", si.size / 1000000, obs_source_get_name(si.source));
+	current_cached_size += si.size;
+	si.cached = true;
 
-	updateSource(si->source, true);
+	updateSource(si.source, true);
 }
 
-void MemoryManager::removeCachedMemory(source_info *si, bool cacheNewFiles)
+// Not thread safe. 'mtx' and 'si.mtx' should be locked
+void MemoryManager::removeCachedMemory(source_info &si, bool cacheNewFiles)
 {
-	std::unique_lock<std::mutex> ulock(si->mtx);
-	if (!si->cached)
+	if (!si.cached)
 		return;
 
-	blog(LOG_INFO, "removing %dMB, source: %s", si->size / 1000000, obs_source_get_name(si->source));
-	current_cached_size -= si->size;
-	si->cached = false;
+	blog(LOG_INFO, "removing %dMB, source: %s", si.size / 1000000, obs_source_get_name(si.source));
+	current_cached_size -= si.size;
+	si.cached = false;
 
-	updateSource(si->source, false);
+	updateSource(si.source, false);
 
 	if (!cacheNewFiles || current_cached_size >= allowed_cached_size)
 		return;
 
-	for (auto data : sources) {
-		if (strcmp(obs_source_get_name(si->source), data.first) != 0 && shouldCacheSource(data.second))
-			addCachedMemory(data.second);
+	for (const auto &data : sources) {
+		std::unique_lock ulock(data.second->mtx);
+		if (strcmp(obs_source_get_name(si.source), data.first) != 0 && shouldCacheSource(*data.second))
+			addCachedMemory(*data.second);
 	}
 }
 
-void MemoryManager::sourceManager(source_info *si)
+void MemoryManager::sourceManager(source_info &si)
 {
-	si->mtx.lock();
-	obs_data_t *settings = obs_source_get_settings(si->source);
+	bool need_get_size = false;
+	{
+		std::unique_lock si_mtx_lock(si.mtx);
 
-	bool looping = obs_data_get_bool(settings, "looping");
-	bool local_file = obs_data_get_bool(settings, "is_local_file");
+		obs_data_t *settings = obs_source_get_settings(si.source);
+		const bool looping = obs_data_get_bool(settings, "looping");
+		const bool local_file = obs_data_get_bool(settings, "is_local_file");
+		obs_data_release(settings);
 
-	obs_data_release(settings);
-	if (!looping || !local_file) {
-		si->mtx.unlock();
-		return;
+		if (!looping || !local_file) {
+			return;
+		}
+
+		need_get_size = si.size == 0;
 	}
-	si->mtx.unlock();
 
-	if (si->size == 0) {
-		uint32_t retry = MAX_POOLS;
+	if (need_get_size) {
+		uint32_t retry = MAX_POLLS;
 		while (retry > 0) {
-			mtx.lock();
-			si->mtx.lock();
-			calculateRawSize(si);
+			si.mtx.lock();
+			calculateRawSize(si); // This also sets 'si.have_video'
 
-			if (si->size || !si->have_video) {
-				si->mtx.unlock();
-				mtx.unlock();
+			if (si.size || !si.have_video) {
+				si.mtx.unlock();
 				break;
 			}
-			si->mtx.unlock();
-			mtx.unlock();
+			si.mtx.unlock();
 
 			retry--;
 			std::this_thread::sleep_for(std::chrono::milliseconds(500));
 		}
 	}
-	std::unique_lock<std::mutex> ulock(mtx);
-	si->mtx.lock();
```

Comment:
[OLD CODE]
Potential dead lock.
Disclaimer: it should not happen in the current code, but... as the code changes some day different threads try to lock mutexes in different order and here we are.